sympy/physics/control/lti.py

```diff
@@ -1422,3 +1424,326 @@ def _eval_rewrite_as_TransferFunction(self, num, den, **kwargs):
+        ========
+
+        >>> from sympy.abc import s, p
+        >>> from sympy.physics.control.lti import TransferFunction, TransferFunctionMatrix
+        >>> G1 = TransferFunction(s + 3, s**2 - 3, s)
+        >>> G2 = TransferFunction(4, s**2, s)
+        >>> G3 = TransferFunction(p**2 + s**2, p - 3, s)
+        >>> tfm1 = TransferFunctionMatrix([[G1], [G2]])
+        >>> tfm1.num_inputs
+        1
+        >>> tfm2 = TransferFunctionMatrix([[G2, -G1, G3], [-G2, -G1, -G3]])
+        >>> tfm2.num_inputs
+        3
+
+        """
+        return self._num_inputs
+
+    @property
+    def num_outputs(self):
+        """
+        Returns the number of outputs of the system.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p, a, b
+        >>> from sympy.physics.control.lti import TransferFunction, Series, TransferFunctionMatrix
+        >>> tf1 = TransferFunction(a*p**2 + b*s, s - p, s)
+        >>> tf2 = TransferFunction(s**3 - 2, s**4 + 5*s + 6, s)
+        >>> tf3 = TransferFunction(a*s - 4, s**4 + 1, s)
+        >>> S1, S2 = Series(tf1, -tf2), Series(tf2, tf3, -tf1)
+        >>> TFM1 = TransferFunctionMatrix([[tf1], [tf2], [tf3]])
+        >>> TFM1.num_outputs
+        3
+        >>> TFM2 = TransferFunctionMatrix([[S1, S2], [-S2, -S1]])
+        >>> TFM2.num_outputs
+        2
+
+        """
+        return self._num_outputs
+
+    @property
+    def shape(self):
+        """
+        Returns the shape of the transfer function matrix, that is, ``(# of outputs, # of inputs)``.
+
+        If explicitly provided as the second argument of ``TransferFunctionMatrix``, then returns the same,
+        else, ``TransferFunctionMatrix`` automatically looks up the shape from the list/tuple provided as
+        the first argument and returns that.
+
+        Examples
+        ========
+
+        >>> from sympy.abc import s, p
+        >>> from sympy.physics.control.lti import TransferFunction, TransferFunctionMatrix
+        >>> tf1 = TransferFunction(p**2 - 1, s**4 + s**3 - p, p)
+        >>> tf2 = TransferFunction(1 - p, p**2 - 3*p + 7, p)
+        >>> tf3 = TransferFunction(3, 4, p)
+        >>> tfm1 = TransferFunctionMatrix([[tf1, -tf2]])
+        >>> tfm1.shape
+        (1, 2)
+        >>> tfm2 = TransferFunctionMatrix([[-tf2, tf3], [tf1, -tf1]])
+        >>> tfm2.shape
+        (2, 2)
+
+        """
+        return self._num_outputs, self._num_inputs
+
+    def __neg__(self):
+        neg = -self._expr_mat
+        return _to_TFM(neg, self.var)
+
+    def __getitem__(self, key):
+        trunc = self._expr_mat.__getitem__(key)
+        if isinstance(trunc, ImmutableMatrix):
+            return _to_TFM(trunc, self.var)
+        else:
+            return TransferFunction.from_rational_expression(trunc, self.var)
+
+    def transpose(self):
+        transposed_mat = self._expr_mat.transpose()
+        return _to_TFM(transposed_mat, self.var)
+
+    def subs(self, dict):
+        return self._expr_mat.subs(dict)
```

Comment:
Basic subclasses should not override subs and should not define it to mean that it returns a different kind of object.